wandb/integration/xgboost/xgboost.py

```diff
@@ -0,0 +1,181 @@
+    - Capture the best eval metric in `wandb.summary` when `define_metric=True` (default).
+
+    Example:
+        ```python
+        bst_params = dict(
+            objective ='reg:squarederror',
+            colsample_bytree = 0.3,
+            learning_rate = 0.1,
+            max_depth = 5,
+            alpha = 10,
+            n_estimators = 10,
+            tree_method = 'hist'
+        )
+
+        xg_reg = xgb.XGBRegressor(**bst_params)
+        xg_reg.fit(X_train, 
+            y_train, 
+            eval_set=[(X_test, y_test)],
+            callbacks=[WandbCallback()])
+        )
+        ```
+    """
+
+    def __init__(
+        self,
+        log_model: bool = False,
+        log_feature_importance: bool = True,
+        importance_type: str = "gain",
+        define_metric: bool = True,
+    ):
+
+        if wandb.run is None:
+            raise wandb.Error("You must call wandb.init() before WandbCallback()")
+
+        self.log_model = log_model
+        self.log_feature_importance = log_feature_importance
+        self.importance_type = importance_type
+        self.define_metric = define_metric
+
+    def before_training(self, model):
+        """Run before training is finished"""
+
+        # Update W&B config
+        config = model.save_config()
+        wandb.config.update(json.loads(config))
+
+        return model
+
+    def after_training(self, model):
+        """Run after training is finished."""
+
+        # Log the booster model as artifacts
+        if self.log_model:
+            self._log_model_as_artifact(model)
+
+        # Plot feature importance
+        if self.log_feature_importance:
+            self._log_feature_importance(model)
+
+        # Log the best score and best iteration
+        if model.attr("best_score") is not None:
+            wandb.log(
+                {
+                    "best_score": float(cast(str, model.attr("best_score"))),
+                    "best_iteration": int(cast(str, model.attr("best_iteration"))),
+                }
+            )
+
+        return model
+
+    def after_iteration(self, model, epoch, evals_log):
+        """Run after each iteration. Return True when training should stop."""
+
+        # Log metrics
+        for data, metric in evals_log.items():
+            for metric_name, log in metric.items():
+                if self.define_metric:
+                    self._define_metric(data, metric_name)
+                    wandb.log({f"{data}-{metric_name}": log[-1]}, commit=False)
+                else:
+                    wandb.log({f"{data}-{metric_name}": log[-1]}, commit=False)
+
+        wandb.log({"epoch": epoch})
+
+        self.define_metric = False
+
+        return False
+
+    def _log_model_as_artifact(self, model):
+        model_name = f"{wandb.run.id}_model.json"
+        model_path = f"{wandb.run.dir}/{model_name}"
+        model.save_model(str(model_path))
+
+        model_artifact = wandb.Artifact(name=model_name, type="model")
```

Comment:
are we doing this anywhere else? i know there are active discussions on how we want to standardize on model logging.

Reply:
I am familiar with Keras WandbCallback and there the best model is saved in the run page itself and not the artifacts. I am confirming this though.

Reply:
Not sure if we are doing it anywhere. (asked AC and Morgan)